### PR TITLE
feat: change propagation algorithm  to particle gun input

### PR DESCRIPTION
**Pull request by asalzburger**
This PR changes that the `PropagationAlgorithm` runs off the particle gun.
Instead of the `PropagationAlgorithm` to generate the input for the propagation, it reads a track parameter collection.

In the default example this is done by:
- particle gun
-  track parameter smearing
- propagation algorithm

Will change reference output.

### CODE REVIEW

File: Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp [hunk 1]
```diff
@@ -67,6 +67,8 @@ class ParticleSmearing final : public IAlgorithm {
         0.1 / Acts::UnitConstants::GeV,  1 * Acts::UnitConstants::ns};
     /// Inflate the initial covariance matrix
     std::array<double, 6> initialVarInflation = {1., 1., 1., 1., 1., 1.};
+    /// Add covariance to the generated parameters
+    bool addCovariances = true;
```
Comment (andiwand) [inline Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp:71]: Thinking about it I believe it would be cleaner to push this flag to the `PropagationAlgorithm` because this is the only case where we want to ignore the covariances I think?

File: Examples/Scripts/Python/propagation.py [hunk 2]
```diff
@@ -18,43 +24,63 @@ def runPropagation(trackingGeometry, field, outputDir, s=None, decorators=[]):
...
+    )
+    s.addAlgorithm(trackParametersGenerator)
+
     nav = acts.Navigator(trackingGeometry=trackingGeometry)
 
     stepper = acts.EigenStepper(field)
     # stepper = acts.AtlasStepper(field)
     # stepper = acts.StraightLineStepper()
-    # stepper = acts.SympyStepper(field)
+    #  stepper = acts.SympyStepper(field)
```
Comment (andiwand) [inline Examples/Scripts/Python/propagation.py:55]: nitpick: unrelated change

File: Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp (hunk 1 above)
Comment (asalzburger) [reply Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp:71]: This is how it started, however, I get a trackParamters object and I would have to clone it without error - in ATLAS we have a method for this, but I don't really like it.

File: Examples/Scripts/Python/propagation.py (hunk 2 above)
Comment (asalzburger) [reply Examples/Scripts/Python/propagation.py:55]: Correct, it needs #3433 anyway, will change.

File: Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp (hunk 1 above)
Comment (andiwand) [reply Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp:71]: Right but now we change the unrelated components which also conflicts with https://github.com/acts-project/acts/pull/3422

```
TrackParameters(parameter.referenceSurface().getSharedPtr(), parameter.parameters(), std::nullopt, parameter.particleHypothesis());
```
with switching behavior for the covariance does not seem to bad to me in the `PropagationAlgorithm`

Comment (asalzburger) [reply Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp:71]: ok, will backe that in then, is fine for me.

Comment (asalzburger) [reply Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.hpp:71]: Done.

File: Examples/Scripts/Python/material_recording.py [hunk 3]
```diff
@@ -70,7 +70,7 @@ def runMaterialRecording(
         detectorConstructionFactory=detectorConstructionFactory,
         randomNumbers=rnd,
         inputParticles=evGen.config.outputParticles,
-        outputMaterialTracks="material_tracks",
+        outputMaterialTracks="material-tracks",
```
Comment (andiwand) [inline Examples/Scripts/Python/material_recording.py:73]:
```suggestion
        outputMaterialTracks="material_tracks",
```

File: Examples/Scripts/Python/material_recording.py [hunk 4]
```diff
@@ -79,7 +79,7 @@ def runMaterialRecording(
         acts.examples.RootMaterialTrackWriter(
             prePostStep=True,
             recalculateTotals=True,
-            inputMaterialTracks="material_tracks",
+            inputMaterialTracks="material-tracks",
```
Comment (andiwand) [inline Examples/Scripts/Python/material_recording.py:82]:
```suggestion
            inputMaterialTracks="material_tracks",
```

**Review by andiwand — APPROVED** [review]

File: Examples/Scripts/Python/material_recording.py (hunk 3 above)
Comment (asalzburger) [reply Examples/Scripts/Python/material_recording.py:73]: I had that initially, but then I need to change it everywhere (including the python tests), so the initially short PR gets more complicated.

I will address that consistently in a separate PR.

Comment (andiwand) [reply Examples/Scripts/Python/material_recording.py:73]: sure no requirement but this change in this pr here no?

File: Examples/Python/tests/test_examples.py [hunk 5]
```diff
@@ -417,9 +417,6 @@ def test_itk_seeding(tmp_path, trk_geo, field, assert_root_hash):
 def test_propagation(tmp_path, trk_geo, field, seq, assert_root_hash):
     from propagation import runPropagation
 
-    obj = tmp_path / "obj"
-    obj.mkdir()
-
```
Comment (andiwand) [inline Examples/Python/tests/test_examples.py:422]: do you think we can keep the checks on the obj?

File: Examples/Scripts/Python/material_validation.py [hunk 6]
```diff
@@ -72,13 +93,23 @@ def runMaterialValidation(
     p.add_argument(
         "-m", "--map", type=str, default="", help="Input file for the material map"
```
Comment (andiwand) [inline Examples/Scripts/Python/material_validation.py:94]: this could be an optional flag to avoid `if len(args.map) > 0` later

File: Examples/Scripts/Python/propagation.py [hunk 7]
```diff
@@ -18,43 +23,65 @@ def runPropagation(trackingGeometry, field, outputDir, s=None, decorators=[]):
...
-            outputDir=outputDir + "/obj",
-        )
-    )
+    # s.addWriter(
+    #    acts.examples.ObjPropagationStepsWriter(
+    #        level=acts.logging.INFO,
+    #        collection="propagation_steps",
+    #        outputDir=outputDir + "/obj",
+    #    )
+    # )
```
Comment (andiwand) [inline Examples/Scripts/Python/propagation.py:75]: should we keep that?

File: Examples/Python/tests/test_examples.py (hunk 5 above)
Comment (asalzburger) [reply Examples/Python/tests/test_examples.py:422]: We check the writer anyway.

I removed the obj writing default because the `propagation.py` would not work out of the box unless it creates the directory mentioned.

File: Examples/Scripts/Python/propagation.py (hunk 7 above)
Comment (asalzburger) [reply Examples/Scripts/Python/propagation.py:75]: See above, I think for checking if the writer works - we should use the `test_writer.py` in my opinion (which we have), the root file is different as we check the hashes.

File: Examples/Scripts/Python/material_validation.py (hunk 6 above)
Comment (asalzburger) [reply Examples/Scripts/Python/material_validation.py:94]: How do I do that?

Comment (andiwand) [reply Examples/Scripts/Python/material_validation.py:94]:
```suggestion
        "-m", "--map", required=True, type=str, help="Input file for the material map"
```
later
```
if args.map is not None:
```

Comment (asalzburger) [reply Examples/Scripts/Python/material_validation.py]: I actually just removed the default, giving no map is also valid, e.g. for the GenericDetector or alike.

**Comment by asalzburger** [comment]
Hoi @andiwand - all comments should be addressed now.

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [d504fdda5243c21ebfd0848901c8db8a4c33ff6f]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177692)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177691)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177690)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177689)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177688)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177687)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177686)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177685)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177684)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177683)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177682)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177681)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177680)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177679)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177678)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177677)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177676)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177675)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177674)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177673)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177672)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177671)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177670)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177669)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177668)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177667)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177666)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177665)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177664)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177663)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177662)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177661)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42177660)||
